compiler/rustc_codegen_rmc/src/compiler_interface.rs

```diff
@@ -164,13 +160,13 @@ impl CodegenBackend for GotocCodegenBackend {
 
             let symtab_filename = base_filename.with_extension("symtab.json");
             debug!("output to {:?}", symtab_filename);
-            let mut out_file = ::std::fs::File::create(&symtab_filename).unwrap();
-            write!(out_file, "{}", result.symtab.to_irep().to_json().pretty().to_string()).unwrap();
+            let out_file = ::std::fs::File::create(&symtab_filename).unwrap();
+            serde_json::to_writer(out_file, &result.symtab.to_irep()).unwrap();
```

Comment:
I was experimenting with `serde` by doing:
```
            let out_file = ::std::fs::File::create(&symtab_filename).unwrap();
            let writer = BufWriter::new(out_file);
            serde_json::to_writer(writer, &result.symtab);
```
The main differences are:
 1. Using a BufWriter to handle writing.
 2. Serializing the symbol table directly instead of the irep representation of the single table.

This has brought memory usage down to 1GB if I have not made any mistakes in measuring.

Not sure if this is related to (2), but conceptually I think it is wrong to transform the symbol table with `to_irep` before writing the JSON file. The serialization of a symbol table is the irep representation itself. Of course, this solution would need writing custom serializers for some objects in the symbol table, but I think it is worth it.

Reply:
For sure, I agree that this is the right path. I used irep because it was the quick and dirty way to unblock me and a fairly simple change. No harm done.